buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalBwcGitPlugin.java

```diff
@@ -0,0 +1,202 @@
+                });
+
+                String checkoutHash = GlobalBuildInfoPlugin.gitInfo(gitExtension.getCheckoutDir().get()).getRevision();
+                logger.lifecycle("Checkout hash for {} is {}", project.getPath(), checkoutHash);
+                writeFile(new File(project.getBuildDir(), "refspec"), checkoutHash);
+            });
+        });
+    }
+
+    public BwcGitExtension getGitExtension() {
+        return gitExtension;
+    }
+
+    /**
+     * We use a time based approach to make the bwc versions built deterministic and compatible with the current hash.
+     * Most of the time we want to test against latest, but when running delayed exhaustive tests or wanting
+     * reproducible builds we want this to be deterministic by using a hash that was the latest when the current
+     * commit was made.
+     * <p>
+     * This approach doesn't work with merge commits as these can introduce commits in the chronological order
+     * after the fact e.x. a merge done today can add commits dated with yesterday so the result will no longer be
+     * deterministic.
+     * <p>
+     * We don't use merge commits, but for additional safety we check that no such commits exist in the time period
+     * we are interested in.
+     * <p>
+     * Timestamps are at seconds resolution. rev-parse --before and --after are inclusive w.r.t the second
+     * passed as input. This means the results might not be deterministic in the current second, but this
+     * should not matter in practice.
+     */
+    private String maybeAlignedRefSpec(Project project, Logger logger, String defaultRefSpec) {
+        if (project.getProviders().systemProperty("bwc.checkout.align").isPresent() == false) {
+            return defaultRefSpec;
+        }
+
+        String timeOfCurrent = execGit(execSpec -> {
+            execSpec.commandLine(asList("git", "show", "--no-patch", "--no-notes", "--pretty='%cD'"));
+            execSpec.workingDir(project.getRootDir());
+        });
+
+        logger.lifecycle("Commit date of current: {}", timeOfCurrent);
+
+        String mergeCommits = execGit(
+            spec -> spec.commandLine(asList("git", "rev-list", defaultRefSpec, "--after", timeOfCurrent, "--merges"))
+        );
+        if (mergeCommits.isEmpty() == false) {
+            throw new IllegalStateException("Found the following merge commits which prevent determining bwc commits: " + mergeCommits);
+        }
+        return execGit(
+            spec -> spec.commandLine(asList("git", "rev-list", defaultRefSpec, "-n", "1", "--before", timeOfCurrent, "--date-order"))
+        );
+    }
+
+    private void writeFile(File file, String content) {
+        try {
+            FileWriter myWriter = new FileWriter(file, false);
+            myWriter.write(content);
+            myWriter.close();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private String execGit(Action<ExecSpec> execSpecConfig) {
```

Comment:
If we need a git specific method, then the spec should set the executable to git? Otherwise how is this different than just a generic exec helper?

Reply:
I removed the git part of it. at the end its just about using the common working dir